### ISSUE TITLE
Can't pull private image for SingleUser

### ISSUE THREAD

**Issue by cometta** [jupyterhub/zero-to-jupyterhub-k8s#1465]
I am using helm 0.9.0-alpha.1.  In pod/continuous-image-puller-2hlk8, it shows
```

 Normal  Scheduled  6m2s   default-scheduler                           Successfully assigned jupyterhub/continuous-image-puller-2hlk8 to aks-nodepool1-37655849-vmssXXXX3x
  Normal  Pulling    6m1s   kubelet, aks-nodepool1-37655849-vmssXXXX3x  Pulling image "<private_repo>.azurecr.io/jupyter-minimal-notebook:latest"
  Normal  Pulled     6m1s   kubelet, aks-nodepool1-37655849-vmssXXXX3x  Successfully pulled image "<private_repo>.azurecr.io/jupyter-minimal-notebook:latest"
  Normal  Created    6m     kubelet, aks-nodepool1-37655849-vmssXXXX3x  Created container image-pull-singleuser
  Normal  Started    6m     kubelet, aks-nodepool1-37655849-vmssXXXX3x  Started container image-pull-singleuser
  Normal
```

but when i use jupyther hub to spin up a user with similar private image <private_repo>.azurecr.io/jupyter-minimal-notebook:latest , i get 

```
  Normal   Pulling                 8s    kubelet, aks-nodepool1-37655849-vmssXXXX3x  Pulling image "<private_repo>.azurecr.io/jupyter-minimal-notebook:latest"
  Warning  Failed                  7s    kubelet, aks-nodepool1-37655849-vmssXXXX3x  Failed to pull image "<private_repo>.azurecr.io/jupyter-minimal-notebook:latest": rpc error: code = Unknown desc = Error response from daemon: Get https://<private_repo>.azurecr.io/v2/jupyter-minimal-notebook/manifests/latest: unauthorized: authentication required
  Warning  Failed                  7s    kubelet, aks-nodepool1-37655849-vmssXXXX3x  Error: ErrImagePull
  Normal   BackOff                 7s    kubelet, aks-nodepool1-37655849-vmssXXXX3x  Back-off pulling image "<private_repo>.azurecr.io/jupyter-minimal-notebook:latest"
  Warning  Failed                  7s    kubelet, aks-nodepool1-37655849-vmssXXXX3x  Error: ImagePullBackOff
```

In config.yml file, i have
```
singleuser:
  imagePullSecret: 
      name : my-existing-secret
```

**Comment by cometta** [comment]
removed pullPolicy: "Always" is the workaround. This is a bug when hub needs to pull the image

**Comment by consideRatio** [comment]
The configuration wasn't correct with regards to how this Helm chart and KubeSpawner is configured, which is different from how you would think they would be configured. This will be fixed in #1794 though, and the correct way to configure this in the past would have been:

```yaml
singleuser:
  image:
    pullSecrets: ["my-existing-secret"]
```

I think this will work also after #1794, but then some more k8s native options would also work.

**Comment by kyprifog** [comment]
I'm still running into this issue when using singleUser profileList, i've tried configuring imagePullSecrets just about anywhere it makes sense to set them and no matter what I do I still get:
```
Failed to pull image "<IMAGE_NAME>": rpc error: code = Unknown desc = Error response from daemon: pull access denied for samtecspg/samtec-customized-jupyter, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
```
I confirmed I can pull the image after logging in with the credentials locally.  I set up the secret according to these instructions:
https://zero-to-jupyterhub.readthedocs.io/en/latest/reference/reference.html?highlight=imagePullSecret#imagepullsecret

**Comment by consideRatio** [comment]
The absolute latest documentation assumes development release of the helm chart, try that. it should be named 0.9.0-n4xx something, see the orange badge in the top of the readme file of the repo.

**Comment by kyprifog** [comment]
Unfortunately upgrading to that version gives me an error in hook-image-awaiter:
```flag provided but not defined: -pod-scheduling-wait-duration```
i'll keep digging...

**Comment by kyprifog** [comment]
So I confirmed that I'm using the most recent version:
```
helm get all data-ops | grep 0.9.0 | head -n 1
        tag: 0.9.0-n358.h11c2aeb8
```
and i've put `imagePullSecrets: ['secret-name']`
and three different levels, under 'jupyterhub:' under 'singleuser:' and under 'kubespawner:' and still no luck.  I'm gonna do some deeper digging.

**Comment by consideRatio** [comment]
The most recent version is found here: https://jupyterhub.github.io/helm-chart/#development-releases-jupyterhub

It is version: 0.9.0-n419.he870fc85 atm, and the new configuration was added to 0.9.0-n399.hb1d96ed5 it seems.

**Comment by kyprifog** [comment]
```
helm get all <release> | grep 0.9.0-n419 | head -n 1
    chart: jupyterhub-0.9.0-n419.he870fc85
```
but I'm still having no luck with 
```
jupyterhub:
   imagePullSecrets: ["image-pull-secret"]
   singleUser:
       profileList:
          -   ....
              imagePullSecrets: ["image-pull-secret"]
```
ill keep looking

**Comment by kyprifog** [comment]
Looking at the k8s manifest of `hook-image-puller` before it fails it doesn't looking like there is any mention of imagePullSecrets (or secrets of any kind) in any of the initContainers it constructs so I feel like I'm either specifying something wrong or the helm chart isn't working as intended.

**Comment by kyprifog** [comment]
```
kubespawner_override:
   image: <IMAGE_NAME>
   imagePullSecrets: ["SECRETNAME"]
```
i thought I had tried this already but for some reason this started working suddenly

**Comment by kyprifog** [comment]
```
traitlets.traitlets.TraitError: The 'image_pull_secrets' trait of a KubeSpawner instance expected a list or a unicode string, not the NoneType None.
```
Now i'm getting this error.  Tried to specify imagePullSecrets for all images in the profileList

**Comment by consideRatio** [comment]
Your configuration

## Valid way to reference an EXISTING k8s Secret for all pods
```yaml
# this is supposed to make you have a chart wide reference to a
# k8s secret named image-pull-secret you have to have created yourself
imagePullSecrets:
  - "my-existing-image-pull-secret"
```

## Valid way to CREATE and reference k8s Secret for all pods
```yaml
# this configuration CREATES a k8s Secret and reference it for all pods in the helm chart.
imagePullSecret:
  create: true
  registry: my-image-registry-reference
  username: my-username-there
  email: some-email@at-some-domain.com
  password: my-password
```

## Valid way to reference an EXISTING k8s Secret for user pods specifically
```yaml
# this is supposed to make you have a reference to a
# k8s secret named image-pull-secret for the user pods
# don't add this AND the chart wide configuration above
singleUser:
  image:
    pullSecrets:
      - "my-existing-image-pull-secret"
```

**Comment by beaniebag** [comment]
I appear to be having the same issue @kyprifog with the latest master version.  I'm using non-standard images of v0.9.1 jupterhub as I wanted to use an alpine base image instead of ubuntu.  

The only edit I've made to master helm chart was removal of the 
\- -pod-scheduling-wait-duration={{ .Values.prePuller.hook.podSchedulingWaitDuration }} line in jupyterhub/templates/image-puller/job.yaml as the version of image puller didn't support it.
The log I receive when logging into jupyterhub and getting a failing spawn is:

```
[E 2020-10-26 09:52:41.088 JupyterHub web:1788] Uncaught exception POST /hub/login?next= (127.0.0.1)
    HTTPServerRequest(protocol='https', host='site', method='POST', uri='/hub/login?next=', version='HTTP/1.1', remote_ip='127.0.0.1')
    Traceback (most recent call last):
      File "/usr/local/lib/python3.9/site-packages/tornado/web.py", line 1703, in _execute
        result = await result
      File "/usr/local/lib/python3.9/site-packages/jupyterhub/handlers/login.py", line 150, in post
        self.redirect(self.get_next_url(user))
      File "/usr/local/lib/python3.9/site-packages/jupyterhub/handlers/base.py", line 649, in get_next_url
        if user.spawner.active:
      File "/usr/local/lib/python3.9/site-packages/jupyterhub/user.py", line 293, in spawner
        return self.spawners['']
      File "/usr/local/lib/python3.9/site-packages/jupyterhub/user.py", line 140, in __getitem__
        self[key] = self.spawner_factory(key)
      File "/usr/local/lib/python3.9/site-packages/jupyterhub/user.py", line 286, in _new_spawner
        spawner = spawner_class(**spawn_kwargs)
      File "/usr/local/lib/python3.9/site-packages/kubespawner/spawner.py", line 148, in __init__
        super().__init__(*args, **kwargs)
      File "/usr/local/lib/python3.9/site-packages/traitlets/config/configurable.py", line 101, in __init__
        self.config = config
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 604, in __set__
        self.set(obj, value)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 593, in set
        obj._notify_trait(self.name, old_value, new_value)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 1217, in _notify_trait
        self.notify_change(Bunch(
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 1227, in notify_change
        return self._notify_observers(change)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 1264, in _notify_observers
        c(event)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 888, in compatible_observer
        return func(self, change)
      File "/usr/local/lib/python3.9/site-packages/traitlets/config/configurable.py", line 205, in _config_changed
        self._load_config(change.new, traits=traits, section_names=section_names)
      File "/usr/local/lib/python3.9/site-packages/traitlets/config/configurable.py", line 187, in _load_config
        warn(msg)
      File "/usr/local/lib/python3.9/contextlib.py", line 124, in __exit__
        next(self.gen)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 1205, in hold_trait_notifications
        raise e
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 1192, in hold_trait_notifications
        self.set_trait(name, value)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 1436, in set_trait
        getattr(cls, name).set(self, value)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 578, in set
        new_value = self._validate(obj, value)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 610, in _validate
        value = self.validate(obj, value)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 1980, in validate
        self.error(obj, value)
      File "/usr/local/lib/python3.9/site-packages/traitlets/traitlets.py", line 690, in error
        raise TraitError(e)
    traitlets.traitlets.TraitError: The 'image_pull_secrets' trait of a KubeSpawner instance expected a list or a unicode string, not the NoneType None.
[E 2020-10-26 09:52:41.093 JupyterHub log:166] {
      "Accept-Language": "en-GB,en;q=0.9",
      "Accept-Encoding": "gzip, deflate, br",
      "Referer": "https://site/hub/login?next=",
      "Sec-Fetch-Dest": "document",
      "Sec-Fetch-User": "?1",
      "Sec-Fetch-Mode": "navigate",
      "Sec-Fetch-Site": "same-origin",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36",
      "Content-Type": "application/x-www-form-urlencoded",
      "Origin": "https://site",
      "Upgrade-Insecure-Requests": "1",
      "Cache-Control": "max-age=0",
      "Content-Length": "38",
      "X-Scheme": "https",
      "X-Original-Uri": "/hub/login?next=",
      "X-Forwarded-Proto": "https,http",
      "X-Forwarded-Port": "443,80",
      "X-Forwarded-Host": "site",
      "X-Forwarded-For": "127.0.0.1,::ffff:10.42.5.0",
      "X-Real-Ip": "127.0.0.1",
      "X-Request-Id": "bdc4ed6439defb5b85003668de31f2f8",
      "Host": "site",
      "Connection": "close"
    }
```
I'm going to do some more testing on my end and will open a new issue if I'm still having this problem.

**Comment by consideRatio** [comment]
@beaniebag you need to rebuild your hub image so it has the latest version of KubeSpawner installed, and you also need to use the latest version of the image-awaiter.

Please keep in mind that whenever you experience an issue with this project after having hardcoded your own images, it may stem from being out of sync with the Helm chart images.

**Comment by kyprifog** [comment]
@beaniebag just FYI I was finally able to get mine working but only with the kubespawner_override option and I had to take my entire helm chart down and back up and had to go back and delete the /charts directories for daskhub (which is what i'm using here:  https://github.com/samtecspg/data-ops ) as well as for data-ops and rerun `helm dep update` in both places.

**Comment by kyprifog** [comment]
https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1465#issuecomment-714739873
@consideRatio I'm having this issue again.   Your comment here doens't make sense, isn't image expecting a string here and not a map?  According to the error I'm getting here thats what I'm seeing.  Whats the full syntax including the name and tag?

**Comment by kyprifog** [comment]
I think the issue is that I'm setting the image at a deeper level under profileList

**Comment by kszyk8** [comment]
Another thing you need to consider if you are new to Kubernetes: 
> The imagePullSecrets field is a list of references to secrets in the same namespace [source](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/)

It means that you must create a secret using `namespace` keyword
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: your-secret-name
  namespace: your-namespace-name
data:
  .dockerconfigjson: paste here a value of base64 encoded data[".dockerconfigjson"]
type: kubernetes.io/dockerconfigjson
```
You can check if secret was created in correct namespace:
```bash
kubectl --namespace your-namespace-name get secret your-secret-name
```

**Comment by kszyk8** [comment]
Another thing you need to consider if you are new to Kubernetes: 
> The imagePullSecrets field is a list of references to secrets in the same namespace [source](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/)

It means that you must create a secret using `namespace` keyword
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: your-secret-name
  namespace: your-namespace-name
data:
  .dockerconfigjson: paste here a value of base64 encoded data[".dockerconfigjson"]
type: kubernetes.io/dockerconfigjson
```
You can check if secret was created in correct namespace:
```bash
kubectl --namespace your-namespace-name get secret your-secret-name
```

**Comment by viniezaboi** [comment]
@kyprifog 
does this method work after all?

`
kubespawner_override:
   image: <IMAGE_NAME>
   imagePullSecrets: ["SECRETNAME"]
`

If so, do you mind telling me which helm chart & jupyter version you're using?
as it does not work in my case.

thanks

**Comment by consideRatio** [comment]
@viniezaboi that won't work, because everything you put under kubespawner_override must be formatted exactly like KubeSpawner the python package wants it.

This configuration is compatible with many versions:

```yaml
singleuser:
  profileList:
    # everything inside here should be snake_case
    - name: some name
      description: some description
      kubespawner_override:
        image: <IMAGE_NAME>:<IMAGE_TAG>
        image_pull_secrets: "SECRETNAME"
```

This configuration is compatible only with more recent versions of z2jh / kubespawner, I don't know exactly where... 

```diff
 singleuser:
   profileList:
     # everything inside here should be snake_case
     - name: some name
       description: some description
       kubespawner_override:
         image: <IMAGE_NAME>:<IMAGE_TAG>
-        image_pull_secrets: "SECRETNAME"
+        image_pull_secrets: ["SECRETNAME"]
```

**Comment by bsikander** [comment]
I had the same issue. What worked for me was to add a reference to the secret in both hub and single user images. Like

```
singleuser:
  image:
    name: '<repo>.azurecr.io/<container>'
    tag: 'v1.1'
    pullSecrets:
      - "my-image-pull-secret"

.....
hub:
  image:
    name: '<repo>.azurecr.io/<hub-custom-cont>'
    tag: 'v1.1'
    pullSecrets:
      - "my-image-pull-secret" 
.....
```

**Comment by consideRatio** [comment]
@bsikander using a modern version of the Helm chart, you can do a chart-wide configuration:

```yaml
imagePullSecrets:
- "my-image-pull-secret"
```
